### DISCUSSION TITLE
ZFS 2.2 container support for Docker?

### DISCUSSION THREAD

**Discussion by cf-sewe** [moby/moby#46654] [Q&A] [answered]
[ZFS 2.2](https://github.com/openzfs/zfs/releases/tag/zfs-2.2.0) implements some PRs for container support:

> Linux container support (https://github.com/openzfs/zfs/pull/12209, https://github.com/openzfs/zfs/pull/14070, https://github.com/openzfs/zfs/pull/14097, https://github.com/openzfs/zfs/pull/12263) - Added support for Linux-specific container interfaces such as renameat(2), support for overlayfs, idmapped mounts in a user namespace, and namespace delegation support for containers.

Some of my questions are:

- Does Docker / Moby needs to be updated to support the features fully?
- Is there are roadmap available for supporting the new features?
- How does https://github.com/moby/moby/pull/45890 relate with the ZFS update?

As we are using Docker with ZFS based container storage this is quite interesting for us.

**Comment by neersighted** [comment, marked as answer]
The critical path you are thinking of is inverted -- the kernel requires these features to use overlayfs with ZFS, and the PR you have linked is intended to make sure that we only try to configure overlayfs on a modern ZFS version that supports the required system calls.

If you have an up-to-date ZFS, any version of Moby with a modern-ish overlay2 should work on that ZFS filesystem. The issues happen when it almost looks like overlayfs is supported (or rather, our existing function tests pass), but those system calls are missing, which is why that PR needs to be completed (adding further function tests to make sure that overlayfs will really work with the backing filesystem).

**Reply by cf-sewe** [reply]
Thank you! I really enjoyed reading through your answer which helped me understand the situation a bit better!
